File: scripts/test_solutions/020_valid_parentheses.py
def isValid(s):
    stack = []
    mapping = {')': '(', '}': '{', ']': '['}
    opens = {'(', '{', '['}
    for c in s:
        if c in mapping:
            if not stack or stack.pop() != mapping[c]:
                return False
        elif c in opens:
            stack.append(c)
    return len(stack) == 0

File: scripts/test_solutions/020_valid_parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> stk;
        unordered_map<char, char> mapping = {
            {')', '('}, {'}', '{'}, {']', '['}
        };
        unordered_set<char> opens = {'(', '{', '['};

        for (char c : s) {
            if (mapping.count(c)) {
                if (stk.empty() || stk.top() != mapping[c]) {
                    return false;
                }
                stk.pop();
            } else if (opens.count(c)) {
                stk.push(c);
            }
        }

        return stk.empty();
    }
};
